DietPi-Update | Add ability to update APT packages

nonnorm opened the pull request:
**Status**: Ready
- [x] Test on VM
- [x] Test on RPi
- [x] Figure out how to have it detect package updates more often than the daily `cron`
- [x] When logging in, it shows up as `You have  package updates available...`

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Update | Add ability to update apt packages and DietPi
- DietPi-Update | CodeFactor Fixes
- DietPi-Update | CodeFactor notice fix
- DietPi-Update | Switch back to echo
- DietPi-Update | Switch to !=
- DietPi-Update | Use (( on if statement
- DietPi-Update | CodeFactor fix
- DietPi-Update | Use [[
- DietPi-Update | Switch to [[ on other if statements
- DietPi-Update | Add more conditions
- DietPi-Update | Remove extra G_AGUG
- DietPi-Update | Remove G_AGUP
- DietPi-Update | Switch to grep -oP
- DietPi-Banner | Switch to != for package number
- DietPi-Banner | Change variable to PACKAGE_NUMBER
- DietPi-Update | Update APT packages when INPUT = 1
- DietPi-Update | Move G_DIETPI-NOTIFY
- DietPi-Update | Simplify if statement
- DietPi-Banner | Add sudo into update message
- DietPi-Banner | Switch to asking for DietPi-Update
- DietPi-Update | Switch from == 1 to != 2
- DietPi-Update | Move package update checking to end
- DietPi-Update | Change to INPUT != 2
- DietPi-Update | Fixes
- DietPi-Update | Remove package flag file
- DietPi-Update | Add pre-patches
- DietPi-Update | Apply pre-patches only if actually running the update
- DietPi-Software | Add automated upgrades into dietpi-software
- DietPi-Software | Show the right name
- DietPi-Software | Auto-install iptables-persistent
- DietPi-Update | Only update APT packages when running dietpi-update manually

nonnorm committed:
Add ability to update apt packages and DietPi

nonnorm committed:
CodeFactor Fixes

nonnorm committed:
CodeFactor notice fix

nonnorm committed:
Switch back to echo

nonnorm committed:
Switch to !=

nonnorm committed:
Use (( on if statement

nonnorm committed:
CodeFactor fix

nonnorm committed:
Use [[

nonnorm committed:
Switch to [[ on other if statements

nonnorm committed:
Add more conditions

nonnorm committed:
Remove extra G_AGUG

nonnorm committed:
Remove G_AGUP

nonnorm committed:
Switch to grep -oP

nonnorm committed:
Switch to != for package number

nonnorm committed:
Change variable to PACKAGE_NUMBER

nonnorm committed:
Update APT packages when INPUT = 1

nonnorm committed:
Move G_DIETPI-NOTIFY

nonnorm committed:
Simplify if statement

nonnorm committed:
Add sudo into update message

nonnorm committed:
Switch to asking for DietPi-Update

nonnorm committed:
Switch from == 1 to != 2

nonnorm committed:
Move package update checking to end

nonnorm commented:
Seems to be ready.  The only problem is that to detect the package upgrades immediately, you have to use `dietpi-update`.  If you just use `apt(-get)`, you need to wait until the next time `cron` runs for the banner to update.  Also, do you think that `1` should be the default, to only show package updates, or `2`, to automatically update packages?  Maybe offer `2` per- software as an option instead of `unattended-upgrades`?
#
Maybe even another option in `dietpi-update`, to just upgrade APT packages (and the banner)?

nonnorm committed:
Change to INPUT != 2

MichaIng commented:
Many thanks, awesome since I wanted to do this for a long time already.

But one thing has changed since my last throughts on this:
- Automatic DietPi updates is not something I would do anymore. We do `dietpi-software` reinstalls and inform users about important changes which both should be recognised interactively, some reinstalls even require interactive input to not hang. So option `3` basically needs to be removed.

> Figure out how to have it detect package updates more often than the daily

I think a daily check is more than enough, don't you? On Debian stable (Buster) new packages are available moreless weekly, I think.

Another issue is that current, APT updates+upgrades are by intention done after pre-patches. We use pre-patches to fix APT preferences or broken repositories or fix other things that would break APT. This definitely needs to stay, while the current PR stage would do APT updates + upgrades before pre-patches.

However, pre-patches of course cannot be done for automated APT upgrades, so I think, if not DietPi update is available, a separate code block or function needs to be called which checks for package updates and in case performs them. Or we need to skip pre-patches if no DietPi update is available.

And not yet 100% sure, but I think we should stop and restart services as well when doing APT upgrades only, especially since some daemons are stopped but not started on upgrade when their state is "disabled", which is true for all DietPi-Services-controlled services.

And a minor thing with the banner, let's do it like this:
- If a DietPi update is available, only show info about this, skip the info about available APT updates.
- Only if no DietPi update is available, check for APT updates (the flag/list file) and show available APT packages in case. And in that case, simply suggest to run `apt upgrade`, or are there benefits do have it wrapped within `dietpi-update`?
___________
Btw, I can also work on the above suggestions, or open a PR against your branch so that you can review first.

nonnorm commented:
What I meant by having it check more than daily, and pretty much the only benefit of running in in `dietpi-update`, is that it automatically updates the banner, so you don't have the message staying there until whenever `cron` runs.

nonnorm committed:
Fixes

MichaIng commented:
Ah, that is true. Hmm, I think we could add an `/etc/apt/apt.conf.d` hook to remove the file or contents from the file when APT installs run. Otherwise we could add the removal at least to `G_AGUG`. I'll have a look into that.

nonnorm commented:
The closest APT hook I can find to what we want is `DPkg::Post-Invoke {"/bin/rm -f /run/dietpi/.package_update";};`.  However, this also runs on installs and uninstalls.

nonnorm committed:
Remove package flag file

nonnorm committed:
Add pre-patches

nonnorm committed:
Change message

nonnorm commented:
@MichaIng, do you think there should be a way in `dietpi-software` to enable automated updates (and ask to be sure) for software that recommends `unattended-upgrades`?

nonnorm committed:
Apply pre-patches only if actually running the update

nonnorm committed:
Add automated upgrades into dietpi-software

nonnorm committed:
Show the right name

nonnorm committed:
Auto-install iptables-persistent

MichaIng commented:
Okay let's get this merged. I'll apply some fine tuning, hope you agree.

MichaIng committed:
v6.34

+ dietpi.txt | Separate settings for DietPi updates and APT updates since we might want to allow automated Python pip, Node,js npm and similar package managers checks/upgrades as well in the future. A single setting then would be too complicated, respectively limits flexibility.
+ dietpi.txt | Minor format/wording changes

MichaIng committed:
v6.34

+ DietPi-Software | Update from CONFIG_CHECK_UPDATES to CONFIG_CHECK_APT_UPDATES option

MichaIng committed:
v6.34

+ DietPi-Cron_daily | Run APT updates and upgrades directly from daily cron job, if daily DietPi update checks are disabled. This allows easier implementation for APT and potentially other package manager update checks/runs without a DietPi update check. Let's leave that flexibility for user choice.

MichaIng committed:
v6.34

+ APT | Remove updates flag file only in "install" action and when it exists

MichaIng committed:
v6.34

+ DietPi-Cron_daily | APT updates flag file is removed by apt-get, hence does not need to be removed manually

MichaIng committed:
v6.34

+ DietPi-Banner | Print info about available APT updates as well in banner header
+ DietPi-Banner | Make DietPi/APT update check functions return result and merge count/version variable with flag variable. This allows a few simplifications.

nonnorm committed:
Only update APT packages when running dietpi-update manually

MichaIng commented:
I'm currently working on the `dietpi-update` file 😉.

MichaIng committed:
v6.34

+ DietPi-Cron_daily | Typo

MichaIng committed:
v6.34

+ DietPi-Update | Restructure: Apply pre-patches only when doing a full DietPi update; Check for APT updates and in case apply them whenever no DietPi update has been found or running in cron mode "2"; Some coding simplifications.

MichaIng committed:
v6.34

+ Tiny

MichaIng committed:
v6.34

+ DietPi-Cron_daily | Check for APT updates by default, i.e. as well in case of invalid or missing dietpi.txt setting, skip only when explicitly "0" is set.

MichaIng commented:
Okay, made a few changes:
- Dedicated dietpi.txt settings to allow more flexibility and e.g. additional pip/npm/... updates in the future.
- Pre-patches must only before actual DietPi updates, hence I separated DietPi updates from APT-only updates code.
- The way how dietpi-update is called from cron job, the APT-only updates/upgrades must run as well when using input "2" (check only). I'm not yet 100% happy with the fact that APT upgrades (if chosen) now run automatically when dietpi-update is called without input (menu mode) but no DietPi update is found. It should be fine as user enabled daily automated upgrades anyway, but it is not consistent with when exiting the menu and probably comes unexpected. I'll think about this later.
- APT installs will now only remove the flag file on the "install" step and when it exists. Else it is called about 5 times or more for each single package install, quite some overhead. Sadly there is no way to restrict it to a single time for an apt call and/or for "apt upgrade" only (instead of single package installs).
- Also the short banner header now shows brief info about available APT updates, same as it also shows info about an available DietPi update.

MichaIng committed:
v6.34

+ DietPi-Banner | Fix flag file name

MichaIng committed:
v6.34

+ DietPi-Banner | Apply colour for APT package update hint

MichaIng committed:
v6.34

+ Tiny

MichaIng committed:
v6.34

+ DietPi-Cron_daily | We need no error handling for APT within the cron job, hence use native apt-get
+ DietPi-Cron_daily | Mute "apt does not have a stable CLI interface" error

MichaIng committed:
v6.34

+ DietPi-Update | Allow G_AGUP/G_AGUG-only failure without error handling as it does not break any further steps at this stage.

MichaIng committed:
v6.34

+ APT | Fix APT update flag file removal: DPKG_HOOK_ACTION variable is only available when invoking via cmd option, not via config option... Never return failure and since "install" action is not done during package installs, use "unpack".

MichaIng committed:
v6.34

+ Tiny

nonnorm commented:
All looks (and works) good to me.  Only thing that I noticed is that there's a typo in the MOTD.

MichaIng commented:
You mean in the APT notice or in the current MOTD (apart from this PR)?

> Take care yourself and the people around you :)

Someone with creativity should forge one, but spelling looks right to me?

nonnorm commented:
Missing an 'of' (in the current MOTD).

> Take care `of` yourself and the people around you :)

MichaIng commented:
Ah that is true, grammatically, however I think "Take care yourself" is still well understood as casual language? I fixed it regardless (requires a reboot or daily cron to be seen in banner).

nonnorm commented:
I think it is understood, but mainly in a different context. Anyway, everything else looks good to me.

MichaIng committed:
v6.34

+ DietPi-Software | Automated APT upgrade prompt: Add info that those are applied on a daily basis and that logs can be found at: /var/tmp/dietpi/logs/dietpi-update_apt.log

MichaIng committed:
v6.34

+ dietpi.txt | Restore info about download size for daily DietPi update check, be more precise as it's 7 bytes
+ dietpi.txt | Add info about logs for daily APT upgrades.

MichaIng commented:
Found probably a better APT hook to remove the flag file: [Pre-Install-Pkgs](https://manpages.debian.org/buster/apt/apt.conf.5.en.html#HOW_APT_CALLS_DPKG(1))
This is executed only once, I'll check if its for installs only.

Sadly there is no `Post-Install-Pkgs` so that the flag stays if an issue occurs.
A flag file for errors would be btw great. But we can think about this at a later time.

nonnorm commented:
Perfect.

MichaIng commented:
Sadly not, as it is invoked for single package removals/purges as well. I.e. it runs before any invocation of dpkg from apt 😞.

... okay, it gets a list of installed DEB packages to STDIN, which is empty on removals. So indeed only when there are DEB packages installed STDIN has content, which we can use.

MichaIng committed:
v6.34

+ APT | Change method for flag file removal: Use Pre-Install-Pkgs which is invoked only a single time. It gets invoked on removals as well but only with any DEB packages are installed, those are sent to STDIN. So we can use read to verify that anything has been sent to STDIN and only then check and remove the flag file. Since this runs directly before invoking dpkg, which does unpacking as very first task, this runs practically at the same step as the previous solution.

MichaIng commented:
Okay will merge now. Makes sense to test it a bit until release, also watch the output when running `dietpi-update` manually, probably we should add some notifications before doing the APT update/upgrade.

Joulinar commented:
Guys I know I'm late and I need to say I did not check the entire code. Therefore my stupid question. What happen if someone is already using `unattended-upgrades` or is going to have it installed. At least software like PiVPN is offering it during install. Will this lead to have 2 competing processes running trying to run apt unattended?

nonnorm commented:
I added a `sed` replace to stop PiVPN from installing `unattended-upgrades`.

MichaIng commented:
Jep, `unattended-upgrades` does anyway not work OOTB since DietPi images are shipped with:
```
2020-11-18 22:29:32 root@VM-Buster:/tmp# l /etc/systemd/system/apt*
lrwxrwxrwx 1 root root 9 Jan  7  2020 /etc/systemd/system/apt-daily-upgrade.service -> /dev/null
lrwxrwxrwx 1 root root 9 Jan  7  2020 /etc/systemd/system/apt-daily-upgrade.timer -> /dev/null
lrwxrwxrwx 1 root root 9 Jan  7  2020 /etc/systemd/system/apt-daily.service -> /dev/null
lrwxrwxrwx 1 root root 9 Jan  7  2020 /etc/systemd/system/apt-daily.timer -> /dev/null
```
to avoid random APT locks and since it is not really a lightweight implementation and AFAIK on Raspbian does not work OOTB since it does not have a security repository.

Joulinar commented:
Ok, it was just something stepping into my mind

Joulinar commented:
Guys, sorry for coming back on this topic. Today I got noticed on on available apt package. However I'm missing the log file `/var/tmp/dietpi/logs/dietpi-update_apt.log` as described on the changelog 😉 

https://github.com/MichaIng/DietPi/blob/b354689141bf3b238b87d9ff83f33df5cf52173f/CHANGELOG.txt#L11

```
 ─────────────────────────────────────────────────────
 DietPi v6.34.3 : 1 APT updates available
 ─────────────────────────────────────────────────────
 - Device model : RPi 4 Model B (armv7l)
 - Uptime : up 4 days, 11 hours, 53 minutes
 - CPU temp : 38'C : 100'F (Cool runnings)
 - LAN IP : 192.168.0.11 (eth0)
 - Info Text : !!! PRODUCTION SYSTEM !!!
 ─────────────────────────────────────────────────────

 DietPi Team     : MichaIng (lead), Daniel Knight (founder), Joulinar (support)
 Image           : DietPi Core Team (pre-image: Raspbian Lite)
 Web             : https://dietpi.com | https://twitter.com/DietPi_
 Donate          : https://dietpi.com/#donate
 DietPi Hosting  : Powered by https://myvirtualserver.com

 apt upgrade     : Run now to apply 1 available APT package upgrades.

root@DietPiProd:~# ls -la /var/tmp/dietpi/logs
total 128
drwxr-xr-x 3 root   root    4096 Oct 23 13:56 .
drw-rw---- 3 dietpi dietpi  4096 Oct 23 13:57 ..
-rw-r--r-- 1 root   root    8588 Aug 28 19:52 dietpi-boot.log
-rw-r--r-- 1 root   root    8843 Sep 26  2019 dietpi-firstboot.log
-rw-r--r-- 1 root   root   60911 Jan 18  2020 dietpi-firstrun-setup.log
-rw-r--r-- 1 root   root     696 Aug 28 19:51 dietpi-preboot.log
-rw-r--r-- 1 root   root    9376 May 19  2020 dietpi-ramdisk.log
-rw-r--r-- 1 root   root     318 Dec 25 00:40 dietpi-ramlog.log
drwxr-xr-t 6 root   root    4096 May 22  2020 dietpi-ramlog_store
-rw-r--r-- 1 root   root    5883 Dec 25 00:32 dietpi-update.log
-rw-r--r-- 1 root   root    1315 Sep 26  2019 fs_partition_resize.log
root@DietPiProd:~#
```

MichaIng commented:
This is only used when you set CONFIG_CHECK_APT_UPDATES to 2 to not only check for updates but apply them as well automatically.

Not sure what a log could for the check itself could contain. I was thinking to store the list of updates, but before doing something like `cat /path/to/log`, `apt list --upgradable` or `apt upgrade` itself can be used to get an overview, so a separate log for this seems nonsense. We could also show the list of packages in the banner instead of the count, but I fear that this is too intrusive for conservative admins as it can easily be multiple lines.

Since "apt upgrade" requires interactive confirmation, we could change to word "apply" to "review" to make careful admins more confident?

Joulinar commented:
ah ok so I misunderstood it 🤣 

> We could also show the list of packages in the banner instead of the count

No I would not do that

> so a separate log for this seems nonsense

You are right. It would be some kind of duplicate data

> Since "apt upgrade" requires interactive confirmation, we could change to word "apply" to "review" to make careful admins more confident?

I guess it's fine as it is now. It was just me having a different expectation 🙄
